### PR TITLE
bpo-33569 Preserve type information with InitVar

### PR DESCRIPTION
**Pull request by hackaugusto**
https://bugs.python.org/issue33569

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by ericvsmith** [comment]
The failing tests need to be addressed. Also, a news blurb is required.

**Comment by hackaugusto** [comment]
@ericvsmith I can't reproduce the build failure, the logs from the build:

```
$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ./python Tools/scripts/patchcheck.py --travis $TRAVIS_PULL_REQUEST; fi
Getting base branch for PR ... origin/master
Getting the list of files that have been added/changed ... 2 files
Fixing Python file whitespace ... 1 file:
  Lib/dataclasses.py
Fixing C file whitespace ... 0 files
Fixing docs whitespace ... 0 files
Please fix the 1 file(s) with whitespace issues
(on UNIX you can run `make patchcheck` to make the fixes)
```

The output from my machine:

```
./python ./Tools/scripts/patchcheck.py
Getting base branch for PR ... origin/master
Getting the list of files that have been added/changed ... fatal: ambiguous argument 'origin/master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
0 files
Fixing Python file whitespace ... 0 files
Fixing C file whitespace ... 0 files
Fixing docs whitespace ... 0 files
Docs modified ... NO
Misc/ACKS updated ... NO
Misc/NEWS.d updated with `blurb` ... NO
configure regenerated ... not needed
pyconfig.h.in regenerated ... not needed
```

I assumed it could be something with local branch, which doesn't have an upstream set, so I refetched master and rebased, but the error persist. Any pointers? Thanks

**Comment by ericvsmith** [comment]
@hackaugusto: I've checked but everything looks okay to me. I can't see what the problem is, unfortunately.

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -198,10 +198,13 @@ def __repr__(self):
...
     def __getitem__(self, params):
-        return self
+        return InitVar(params)
 
 class InitVar(metaclass=_InitVarMeta):
-    pass
+    __slots__ = ('type', )
+
+    def __init__(self, type):
+         self.type = type
```
Comment (ericvsmith) [inline Lib/dataclasses.py]: Found it: this line has 5 space indentation.

Comment (hackaugusto) [reply Lib/dataclasses.py]: ohhh, little bastard. thanks, should be fixed now :)

File: Lib/dataclasses.py [hunk 2]
```diff
@@ -198,10 +198,13 @@ def __repr__(self):
 
 class _InitVarMeta(type):
     def __getitem__(self, params):
-        return self
+        return InitVar(params)
 
 class InitVar(metaclass=_InitVarMeta):
-    pass
+    __slots__ = ('type', )
```
Comment (ericvsmith) [inline Lib/dataclasses.py:207]: We should probably have a nicer `__repr__` here.

Currently you get:
```
>>> from dataclasses import InitVar
>>> InitVar[int]
<dataclasses.InitVar object at 0x00A041C0>
```

I'd suggest something like:
```
    def __repr__(self):
        return f'dataclasses.InitVar[{self.type}]'
```
for starters, although `typing.ClassVar` does something more sophisticated in place of `{self.type}`. Maybe someday when we import `typing` from `dataclasses` we can switch to using their helper function. It's probably worth adding a comment to that effect, too.

Comment (hackaugusto) [reply Lib/dataclasses.py:207]: Using `f'dataclasses.InitVar[{self.type}]'` would look like:

```python
>>> from dataclasses import InitVar
>>> InitVar[int]
dataclasses.InitVar[<class 'int'>]
```

So I added `f'dataclasses.InitVar[{self.type.__name__}]'` instead. Should I write a test for this?

Comment (ericvsmith) [reply Lib/dataclasses.py:207]: If that's what gives you the equivalent of:
```
>>> import typing
>>> typing.ClassVar[int]
typing.ClassVar[int]
```
Then I'm all for it.

Yes, we'll need tests for this. Nothing complex, `InitVar[int]` should be good enough.

Comment (hackaugusto) [reply Lib/dataclasses.py:207]: Sorry for the delay, I added a really simple test:

`self.assertEqual(InitVar[int].type, int)`

Would you like tests using `__repr__` and `@dataclass`?

Comment (ericvsmith) [reply Lib/dataclasses.py:207]: I apologize for taking forever to get back to you on this.

Could you test the `repr` of an `InitVar`? Something like (untested) `repr(InitVar[int]) == 'dataclasses.InitVar[int]'`?

I promise I'll commit this as soon as you add this change.

**Pull request merged**